### ISSUE TITLE
securityContext.fsGroup setting lead to permission issue on restart / remount

### ISSUE THREAD

**Issue by marcoschulte** [CrunchyData/postgres-operator#3023]
## Overview

Our postgres instances regularly stop working with following in the logs:

```
2022-02-08 13:36:54.275 UTC [1257059] FATAL:  data directory "/pgdata/pg13" has invalid permissions
2022-02-08 13:36:54.275 UTC [1257059] DETAIL:  Permissions should be u=rwx (0700) or u=rwx,g=rx (0750).
```

Indeed, after restarts / remounts, the pg13 folder has permissions `rwxrwx---`, even though we removed group write manually.

We suspect the issue is that `securityContext.fsGroup` is being set at https://github.com/CrunchyData/postgres-operator/blob/2e18aef93dd2d6dee065ad00c959dc9fabc6da79/internal/postgres/reconcile.go#L280

According to https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/#security-context this causes the csi driver to add the group `26` to the volume **and** set permissions to current permissions OR `rw-rw----`, therefore always adding the unwanted g+w.

We tried setting `openshift: true`, which removes `fsGroup` from the descriptor, but this causes the postgres startup container to fail, as the secrets cannot be read anymore

```
Initializing ...
::postgres-operator: uid::26
::postgres-operator: gid::26
::postgres-operator: postgres path::/usr/pgsql-13/bin/postgres
::postgres-operator: postgres version::postgres (PostgreSQL) 13.5
::postgres-operator: config directory::/pgdata/pg13
::postgres-operator: data directory::/pgdata/pg13
install: cannot open '/pgconf/tls/replication/tls.crt' for reading: Permission denied
install: cannot open '/pgconf/tls/replication/tls.key' for reading: Permission denied
install: cannot open '/pgconf/tls/replication/ca.crt' for reading: Permission denied
```

## Environment

Please provide the following details:

- Platform: k3s v1.22.5+k3s1
- PGO: 5.0.4
- Postgres Version: crunchy-postgres:centos8-13.5-0
- Storage: Hetzner CSI Driver v1.6.0 (https://github.com/hetznercloud/csi-driver)

**Comment by andrewlecuyer** [comment]
@marcoschulte thank you for the detailed submission.

I was hoping to clarify one detail.  When you say the following:

> Indeed, after restarts / remounts, the pg13 folder has permissions rwxrwx---

What is actually being restarted?  E.g. is the entire node being restarted?

**Comment by Azenet** [comment]
Having this happen here as well, single-node kubeadm test cluster with openebs-zfs localpv provisioner and PGO 5.0.4. I'm unsure what triggers it on my setup, it doesn't always happen on startup but seems to happen randomly at some point after it has been running for a while. It has happened in the past when the k8s api had become unavailable for a bit, but as far as I can tell the operator had not updated the pod/statefulset when that happened.

**Comment by Remp69** [comment]
Same issue for me. Seems to happened randomly, sometime in one pod, sometime in many pods at the same time.
PGO version is 4.7.0
K8s v1.20
Storage: Hitachi CSI driver v3.7.0

**Comment by marcoschulte** [comment]
> @marcoschulte thank you for the detailed submission.
> 
> I was hoping to clarify one detail. When you say the following:
> 
> > Indeed, after restarts / remounts, the pg13 folder has permissions rwxrwx---
> 
> What is actually being restarted? E.g. is the entire node being restarted?

We tried to reproduce this issue but unfortunately aren't sure anymore about the reasons. It still happens regularly, but seems to be random.

For the ones with this issue, we are using this hack for now: We added following snippet on one kubernetes node to the crontab and run it every minute:

`/usr/local/bin/kubectl get pods -A | grep "\-instance1\-" | awk  '{system("/usr/local/bin/kubectl exec -n " $1 " " $2 " --container database -- chmod 750 /pgdata/pg13")}'`

**Comment by digitalcraig** [comment]
Also seeing this same issue with k3s v1.23.4+k3s1 and Longhorn 1.2.4.

```
2022-04-04 13:24:08,471 INFO: doing crash recovery in a single user mode
2022-04-04 13:24:08,479 ERROR: Crash recovery finished with code=1
2022-04-04 13:24:08,479 INFO:  stdout=
2022-04-04 13:24:08,479 INFO:  stderr=2022-04-04 13:24:08.479 UTC [15883] FATAL:  data directory "/pgdata/pg13" has invalid permissions
2022-04-04 13:24:08.479 UTC [15883] DETAIL:  Permissions should be u=rwx (0700) or u=rwx,g=rx (0750).
```

I'm running 3 instances of 1 replica each and it seems to be the secondaries which fail to completely start (2/3 containers running) nightly after the full backup. Even though there is a pod fully started and it is the leader, the database is not available. Most of the time, I can manually delete the pods and they recover. Sometimes I have to manually start a shell in the running container and fix the permissions. In my case, /pgdata/pg13 has setguid (+s) bit set where it should be 0750.

**Comment by budimanjojo** [comment]
Is there any findings on this? I have this problem almost daily, I don't know if a full backup is causing this like @digitalcraig mentioned but it might be. Is there any workaround for this?

**Comment by 9numbernine9** [comment]
I've run into this issue recently as well. This occurred when doing a full restore from an S3 backup, i.e. creating a brand new cluster with new volumes.

- Platform: k3s v1.23.6+k3s1
- PGO: 5.0.5
- Postgres Version: crunchy-postgres:centos8-14.1-0
- Storage: Longhorn 1.2.4

**Comment by chobostar** [comment]
We are currently experiencing the same issue.

I wrote here some recommendation to make Patroni more tolerant to KubeAPI issues, that will prevent false positive restarts:
https://github.com/CrunchyData/postgres-operator/issues/2838#issuecomment-1125632868

However sure it's not solution for PGDATA permissions.

**Comment by heilerich** [comment]
I don't think there is a bug here on the kubernetes/storage driver side. They are exactly doing what they are supposed to (see the kubernetes documentation linked in the issue's description).

If a fsGroup is set, kubernetes will recursively change the directory permissions on mount. This is going well exactly once: on first startup when the PV is still empty (because PG does not care about the permission on the parent/root directory). If the pod is recreated for some reason, the same thing happens again (as per spec). Only this time the pg13/14 directory will be there as well and its permissions will get changed as well. Which will result in the crashes we are experiencing.

The reason this happens only in some environments, is that not every CSI driver is compatible with the fsGroup specification. So if the CSI driver ignores the fsGroup setting everything will be well. If it actually enforces it the crashes occur.

The fix should be easy: If we do not want the filesystem permissions to change (and we obviously don't because PG does not like it) the fsGroup parameter should be removed from the pod's securityContext. There is no use for it in this case anyways.

**Comment by eliranw** [comment]
+1

**Comment by TomerFisher** [comment]
happen to me too...

**Comment by cbandy** [comment]
We haven't been able to reproduce or pinpoint the cause of these changes to file permissions. It seems likely that kubelet is touching them as part of setting up mounts [according to `fsGroup`](https://kubernetes.io/blog/2020/12/14/kubernetes-release-1.20-fsgroupchangepolicy-fsgrouppolicy/).

If anyone here is willing to try some changes in their affected non-production environment, that would be great! Anyone using Kubernetes 1.23 (or 1.20 with [ConfigurableFSGroupPolicy](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/) enabled) can try the following patch to see if it helps. Applying it will redeploy your Postgres pods.

List your Postgres StatefulSets with:

```sh
kubectl get sts.apps --selector '
  postgres-operator.crunchydata.com/cluster,
  postgres-operator.crunchydata.com/instance
'
```

Apply the patch to each separately:

```sh
kubectl patch sts.apps hippo-instance1-sgzm --patch '
  {"spec":{"template":{"spec":{"securityContext":{"fsGroupChangePolicy":"OnRootMismatch"}}}}}
'
```

**Comment by matzik12** [comment]
Hi,
This solution worked our for me but I am looking for a solution that I can use in the CR. I have a lot of ppstgres clusters and I dont want to change each statefull set every time I create a new cluster. Hope you got an answer for me.
Thanks @cbandy !

**Comment by heilerich** [comment]
This solution will still lead to broken volumes when the ownership of the volume root changes for some reason, because kubelet will start changing the permissions recursively again if that happens.

Admittedly, that should not really happen, but I still don't understand why the operator should set an `fsGroup` at all if we do not want kubelet to touch the volume's permissions. Can somebody explain?

**Comment by cbandy** [comment]
> Admittedly, that should not really happen, but I still don't understand why the operator should set an `fsGroup` at all if we do not want kubelet to touch the volume's permissions. Can somebody explain?

Without it, volumes are mounted such that only `root` (UID 0) may write to them. PGO runs all its containers with `runAsNonRoot: true`, so there's no opportunity from within to change file ownership. The `fsGroup` field instructs the privileged stuff outside the container to make it writable by the processes inside the container.

OpenShift takes this a step further and assigns an (effectively) random UID to each Pod. The `fsGroup` ensures that files are writable even if the Pod is recreated with a new UID. Kubernetes is considering a similar effect with [user namespaces](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/127-user-namespaces).

**tl;dr** it is more secure.

**Comment by matzik12** [comment]
The pgo changes the fs permissions to 700 when init container runs at pod startup. When my postgres (because of any reason) crashes it restarts the postgres container and than the permissions are not changing to 700 because the init container didn't run at the moment. Can't we somehow make the whole pod crash if only on container in it crashes and than the init container will change the permissions?
@cbandy

**Comment by cbandy** [comment]
@matzik12 if you have a reliable way to trigger this bug, please try the patch I suggested. We can incorporate it into PGO if it resolves this issue.

**Comment by heilerich** [comment]
I have been using the change mentioned by @cbandy for about ~6 weeks now and it haven't had problems since then. 

@matzik12 We did not change anything manually in our clusters, but used a policy agent to automatically patch the pods created by the controller manager. Here is an example for [kyverno](https://kyverno.io/), but since OPA Gatekeeper has support for [Mutations](https://open-policy-agent.github.io/gatekeeper/website/docs/mutation) now that should work as well.

```yaml
apiVersion: kyverno.io/v1
kind: ClusterPolicy
metadata:
  name: fix-postgres-fsgroup
spec:
  rules:
  - name: fix-postgres-fsgroup
    match:
      all:
      - resources:
          kinds:
          - Pod
          selector:
            matchExpressions:
            - key: postgres-operator.crunchydata.com/instance
              operator: Exists
    mutate:
      patchStrategicMerge:
        spec:
          securityContext:
            fsGroupChangePolicy: "OnRootMismatch"
```

**Comment by matzik12** [comment]
@cbandy your solution fixes this bug. If you can update it into pgo it will help :)

**Comment by benjaminjb** [comment]
Hello, just wanted to let y'all know if you haven't seen the attached PR, the `fsGroupChangePolicy` fix that was suggested and tested in this thread has been merged in to PGO and should be out in the next release. Reminder (for future people who find this) that this fix is only effective in K8s 1.20+

Hope this helps and if you're still experiencing a problem like this, please let the community know.

**Comment by matzik12** [comment]
@benjaminjb Hi, where can I get those new images when it release? I saw that almost all images that were on crunchy website disappeared...
